components/brave_wallet/browser/ens_resolver_task.cc

```diff
@@ -0,0 +1,495 @@
+    return;
+  }
+
+  auto bytes_result = ParseDecodedBytesResult(body);
+  if (!bytes_result) {
+    ParseErrorResult<mojom::ProviderError>(body, &error_.emplace(),
+                                           &error_message_.emplace());
+    return;
+  }
+
+  // Decoding as returned bytes[] per
+  // https://github.com/ensdomains/docs/blob/e4da40003943dd25fdf7d4c5552335330a9ee915/ens-improvement-proposals/ensip-10-wildcard-resolution.md?plain=1#L70
+  auto decoded_resolve_result =
+      eth_abi::ExtractBytesFromTuple(*bytes_result, 0);
+  if (!decoded_resolve_result) {
+    error_ = mojom::ProviderError::kInternalError;
+    error_message_ = l10n_util::GetStringUTF8(IDS_WALLET_INTERNAL_ERROR);
+    return;
+  }
+
+  resolve_result_ = std::move(*decoded_resolve_result);
+}
+
+void EnsResolverTask::FetchOffchainData() {
+  DCHECK(offchain_lookup_data_);
+
+  GURL offchain_url;
+  bool data_substitued = false;
+  bool valid_sender = true;
+
+  // Sender must match resolver per
+  // https://eips.ethereum.org/EIPS/eip-3668#client-lookup-protocol #5.
+  if (base::CompareCaseInsensitiveASCII(offchain_lookup_data_->sender.ToHex(),
+                                        *resolver_)) {
+    valid_sender = false;
+  }
+
+  // Pick first valid url.
+  // TODO(apaymyshev): Implement picking different url per
+  // https://eips.ethereum.org/EIPS/eip-3668#client-lookup-protocol #9.
+  for (auto url_string : offchain_lookup_data_->urls) {
+    base::ReplaceSubstringsAfterOffset(&url_string, 0, "{sender}",
+                                       offchain_lookup_data_->sender.ToHex());
+    data_substitued = base::Contains(url_string, "{data}");
+    base::ReplaceSubstringsAfterOffset(&url_string, 0, "{data}",
+                                       ToHex(offchain_lookup_data_->call_data));
+    GURL url(url_string);
+    if (url.is_valid() && url.SchemeIsHTTPOrHTTPS() &&
+        url.SchemeIsCryptographic()) {
+      offchain_url = std::move(url);
+      break;
+    }
+  }
+
+  if (!valid_sender || !offchain_url.is_valid()) {
+    error_ = mojom::ProviderError::kInternalError;
+    error_message_ = l10n_util::GetStringUTF8(IDS_WALLET_INTERNAL_ERROR);
+    base::SequencedTaskRunnerHandle::Get()->PostTask(
+        FROM_HERE, base::BindOnce(&EnsResolverTask::WorkOnTask,
+                                  weak_ptr_factory_.GetWeakPtr()));
+    return;
+  }
+
+  std::string payload;
+  if (!data_substitued) {
+    base::Value::Dict payload_dict;
+    payload_dict.Set("sender", offchain_lookup_data_->sender.ToHex());
+    payload_dict.Set("data", ToHex(offchain_lookup_data_->call_data));
+    base::JSONWriter::Write(payload_dict, &payload);
+  }
+
+  url_loader_ = CreateLoader(offchain_url, payload);
+  url_loader_->DownloadToStringOfUnboundedSizeUntilCrashAndDie(
+      url_loader_factory_.get(),
+      base::BindOnce(&EnsResolverTask::OnFetchOffchainDone,
+                     weak_ptr_factory_.GetWeakPtr()));
```

Comment:
I see `JsonReader::Read` will be called by `ExtractGatewayResult` in `OnFetchOffchainDone`. Isn't this response body unsanitized if we use url loader directly instead of our api_request_helper? Also, what's the reason to use url loader directly rather than using api_request_helper?

Reply:
Made it so ENSResolver uses ApiRequestHelper now